api/OBJECT-SIGNING.md

```diff
@@ -0,0 +1,63 @@
+
+## Certificate management
+
+When TLS MiTM is in place the device will trust the proxy at the TLS level either because it is trusted by being in the standard Linux root certificate list, or because it was specified in the `proxyCertPEM` field in the device configuration API. Thus TLS still provides server authentication and privacy plus integrity.
+
+But since there is no TLS client authentication the `AuthContainer` message is used to sign the payloads from the device to the controller (using standard ECDA signing operations (implemented in ecdsa golang package alternatively the Sign function from github.com/google/go-tpm/tpm2 when the device private key is in a TPM).
+
+The device indicates who it is by providing a hash of its of the certificate corresponding to its signing key in the `AuthContainer`, which is used on the controller side to look up the device (and verify the signature using standard ECDSA verification.)
+Since the device never changes its device certificate per definition (it becomes a new device should it ever change it), there is no need for updating the cert in the device to controller direction.
+
+The logic for the signing and verification is the same in the controller to device direction. However, the certificate handling is different.
+The device will use the `ControllerCerts` API to fetch the signing certificates and any intermediate certificates. The certificate chain will be verified using the standard go X.509's package VerifyChain up to the trusted root CA configured at birth of the device (In EVE this is kept in /config/root-certificate.pem and is orthogonal to the set of trusted root CAs used for TLS).
+The fetch using `ControllerCerts` is required at install time and when the `senderCertHash` does not match a certificate already known to the device. Thus the controller can roll certificates as long as it makes them available using the `ControllerCerts` API and puts the hash of the user signer's certificate in the `senderCertHash` in the `AuthContainer`; that will make the device re-issue a GET of `ControllerCerts`.
+
+During the initial device onboarding the `register` API might be used. Since the controller might not know all of the onboarding certificates a-priori the device MUST include the full `senderCert` in the `AuthContainer`.
+
+## Signing and verification
+
+With the certificates in place the message payloads can be signed and verified on both directions. The 'AuthContainer' has the following fields:
+
+- `authPayload`: marshalled protobuf message containing actual payload
+- `algo`: hash algorithm used in the 'AuthContainer', currently it's either a SHA256-32bytes or SHA256-16bytes
+- `senderCertHash`: the hash value of the sender's certificate used
+- `signatureHash`: the signed hash of the marshalled protobuf message
+- `sendCert`: device MUST send the 'onboarding' certificate in the `register` message
```

Comment:
Why? I don't need to send a certificate, only prove that I have a valid one by signing the message, which I do.

Reply:
Somewhere else the document says that for onboarding the controller might not have all of the onboarding certificates known a-priori (but have some way to validate the onboarding certificates)

Reply:
Ah

* if the controller has the onboard cert, then signing is enough
* if the controller has the CA that signed the onboard cert, but not the onboard cert itself, then it needs the actual cert along with signature to validate

That makes sense, thanks. And (as I argued above), simplicity says to just send it every onboard time, not try to figure out when to send and when not.

👍

Reply:
Except that it would make every ConfigRequest 10-100x larger since the certs are huge and the ConfigRequest/ConfigResponse is only a hash of the latest config by default. Bandwidth isn't free at the edge.
That is also why we want to be able to have a shorter sha in the senderCertHash.